### DISCUSSION TITLE
Showing a way of returning elementwise operations to some functions

### DISCUSSION THREAD

**Discussion by dvd101x** [josdejong/mathjs#2881] [Show and tell]
Hi everyone,

I have been working on a way to extend some mathjs functions to add elementwise operations similar to matlab or numpy (personally I'm more familiar with that)

> trigonometric functions, exp, log, gamma, square, sqrt, cube, and cbrt

To test it:

* Open this site https://dvd101x.github.io/web-thermodynamics/
* Paste these commands on the left pane.

```python
x = 1:3

exp(x)
log(x)
log(x,2)
gamma(x)
square(x)
cube(x)
cbrt(x)

# Trigonometrics
acos(x)
acosh(x)
acot(x)
acoth(x)
acsc(x)
acsch(x)
asec(x)
asech(x) 
asin(x)
asinh(x) 	
atan(x)
atan2(x, x)
atanh(x)
cos(x)
cosh(x)
cot(x)
coth(x)
csc(x)
csch(x)
sec(x)
sech(x)
sin(x)
sinh(x)
tan(x)
tanh(x)
```

It's the first time I use some of the extension capabilities of mathjs and I'm fascinated by that. Maybe it's a bit basic for some of the more advanced users here and I see those capabilities are well documented, nontheless here it is.

Even found a way to make work `log([1, 2], 2)` **logarithm base 2 elementwise for [1, 2]**

https://github.com/dvd101x/web-thermodynamics

PD: Next I will try to extend index

**Comment by josdejong** [comment]
Thanks for sharing David!

You can indeed extend a mathjs function with support for `Array` and `Matrix`, for example to extend the function `sin`:

```js
// extend function sin with support for matrices and arrays
math.import({
  sin: math.typed({
    'Array | Matrix': A => math.map(A, math.sin)
  })
})
```

For reference: there is a discussion whether to support element-wise operations for a set of functions in #2440

**Reply by dvd101x** [reply]
Thank you Jos!

I will reade more about `math.typed( )` as I currently don't undertand it but seems to be very useful for this and other stuff I'm testing.

**Reply by josdejong** [reply]
👍  you can read more about it on the following page: https://mathjs.org/docs/core/extension.html#typed-functions

**Reply by dvd101x** [reply]
Thank you!

I’m starting to understand it and I’m making a few tests. I will try to update my code accordingly.

**Reply by dvd101x** [reply]
I used it extensively and it worked like a charm.

Thank you so much!
